Feature: conda package for PAM

brynpickering opened the pull request:
With this PR, it is now possible to build a conda package of PAM.

Locally, it requires creating a new conda environment, etc: 

```
mamba create -n conda-builder boa conda-verify
conda config --add channels city-modelling-lab 
mamba activate conda-builder
conda mambabuild .
mamba install -c local --strict-channel-priority pam
```

However, this is more useful for providing the package on the city modelling lab [anaconda channel](https://anaconda.org/city-modelling-lab/), which will be automatically done on each new PAM release (thanks to the new GH actions).

I have also taken the opportunity to move from `setup.py` to `pyproject.toml`, which is the new [preferred method](https://packaging.python.org/en/latest/tutorials/packaging-projects/) for providing the metadata for a package. It is still possible to use `pip install ...` with this new setup.

## Checklist

Any checks which are not relevant to the PR can be pre-checked by the PR creator. 
All others should be checked by the reviewer(s). 
You can add extra checklist items here if required by the PR.

- [ ] CHANGELOG updated
- [ ] Memory profiling tests pass (see [here](https://arup-group.github.io/pam/contributing/coding/#memory-profiling) for more details)
- [ ] Tests added to cover contribution
- [ ] Documentation updated

brynpickering commented:
TODO: get the anaconda token for city-modelling-lab and add it to repo and AWS secrets. Not strictly necessary until the next PAM release, but maybe we release one soon to get our first PAM package online?

brynpickering commented:
Example of GH action to pre-build the conda package (on creating and pushing a version tag): https://github.com/brynpickering/pam/actions/runs/5489256916

Example GH action to publish a new docs version to the GH pages branch and to publish the built conda package to anaconda: https://github.com/brynpickering/pam/actions/runs/5489472569

Built package currently going here (as the anaconda token links to my channel): https://anaconda.org/brynpickering/pam

Would be good if you can try creating a conda package and installing this into it @mfitz @fredshone : `mamba create -n pam-test -c brynpickering pam`

mfitz reviewed (commented):
When I built the package locally it seemed to be happy about building, but not about subsequent tests @brynpickering (although that looks to be a code coverage config issue):

```bash
## Package Plan ##

  environment location: /Users/mickyfitz/micromamba/envs/conda-builder/conda-bld/pam_1689270940568/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol


The following NEW packages will be INSTALLED:

    anyio:                         3.7.1-pyhd8ed1ab_0         conda-forge
    appnope:                       0.1.3-pyhd8ed1ab_0         conda-forge
     ...
    zstd:                          1.5.2-h4f39d0f_7           conda-forge

Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
export PREFIX=/Users/mickyfitz/micromamba/envs/conda-builder/conda-bld/pam_1689270940568/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol
export SRC_DIR=/Users/mickyfitz/micromamba/envs/conda-builder/conda-bld/pam_1689270940568/test_tmp
+ pytest tests/
============================= test session starts ==============================
platform darwin -- Python 3.11.4, pytest-7.4.0, pluggy-1.2.0
rootdir: $SRC_DIR
configfile: pyproject.toml
plugins: nbmake-1.3.4, timeout-2.1.0, cov-4.1.0, mock-3.11.1, anyio-3.7.1, xdist-3.3.1
created: 2/2 workers
2 workers [796 items]
...

---------- coverage: platform darwin, python 3.11.4-final-0 ----------

FAIL Required test coverage of 89.0% not reached. Total coverage: 0.00%
=========================== short test summary info ============================
SKIPPED [1] tests/test_03_read_challenge.py:59: Redundant test - we rarely see this requirment any more
SKIPPED [1] tests/test_03_read_challenge.py:66: Redundant test - we rarely see this requirment any more
SKIPPED [1] tests/test_03_read_challenge.py:89: Redundant test - we rarely see this requirment any more
SKIPPED [1] tests/test_03_read_challenge.py:98: Redundant test - we rarely see this requirment any more
SKIPPED [1] tests/test_03_read_challenge.py:107: Redundant test - we rarely see this requirment any more
SKIPPED [1] tests/test_03_read_challenge.py:124: Redundant test - we rarely see this requirment any more
XFAIL tests/test_05_modifiers_ReduceSharedActivity.py::test_remove_activities_removes_relevant_shared_activities
================== 789 passed, 6 skipped, 1 xfailed in 34.42s ==================
Tests failed for pam-0.2.5.dev0-py_0.tar.bz2 - moving package to /Users/mickyfitz/micromamba/envs/conda-builder/conda-bld/broken
```

I'm not sure what success looks like, but these are the apparent working directories:

```bash
% ls -talh /Users/mickyfitz/micromamba/envs/conda-builder/conda-bld/pam_1689270940568/

total 664
drwxr-xr-x@ 11 mickyfitz  staff   352B 13 Jul 18:58 ..
drwxr-xr-x@ 12 mickyfitz  staff   384B 13 Jul 18:58 test_tmp
drwxr-xr-x@ 15 mickyfitz  staff   480B 13 Jul 18:57 _test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol
drwxr-xr-x@  9 mickyfitz  staff   288B 13 Jul 18:57 .
drwxr-xr-x@  2 mickyfitz  staff    64B 13 Jul 18:56 work
drwxr-xr-x@ 32 mickyfitz  staff   1.0K 13 Jul 18:56 work_moved_pam-0.2.5.dev0-py_0_noarch
drwxr-xr-x@  3 mickyfitz  staff    96B 13 Jul 18:56 build_prefix_moved_pam-0.2.5.dev0-py_0_osx-arm64
drwxr-xr-x@  2 mickyfitz  staff    64B 13 Jul 18:56 pip_cache
-rw-r--r--@  1 mickyfitz  staff   330K 13 Jul 18:56 prefix_files.txt

% ls -talh /Users/mickyfitz/micromamba/envs/conda-builder/conda-bld/pam_1689270940568/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol
total 0
drwxr-xr-x@  15 mickyfitz  staff   480B 13 Jul 18:57 .
drwxr-xr-x@ 306 mickyfitz  staff   9.6K 13 Jul 18:57 conda-meta
drwxr-xr-x@ 465 mickyfitz  staff    15K 13 Jul 18:57 bin
drwxr-xr-x@  45 mickyfitz  staff   1.4K 13 Jul 18:57 share
drwxr-xr-x@ 631 mickyfitz  staff    20K 13 Jul 18:57 lib
drwxr-xr-x@ 545 mickyfitz  staff    17K 13 Jul 18:57 include
drwxr-xr-x@  10 mickyfitz  staff   320B 13 Jul 18:57 doc
drwxr-xr-x@   5 mickyfitz  staff   160B 13 Jul 18:57 etc
drwxr-xr-x@   3 mickyfitz  staff    96B 13 Jul 18:57 var
drwxr-xr-x@  14 mickyfitz  staff   448B 13 Jul 18:57 sbin
drwxr-xr-x@   6 mickyfitz  staff   192B 13 Jul 18:57 cmake
drwxr-xr-x@   4 mickyfitz  staff   128B 13 Jul 18:57 man
drwxr-xr-x@   9 mickyfitz  staff   288B 13 Jul 18:57 ssl
drwxr-xr-x@  33 mickyfitz  staff   1.0K 13 Jul 18:57 fonts
drwxr-xr-x@   9 mickyfitz  staff   288B 13 Jul 18:57 ..
```

Does that look right?

After the test stage failure, the archive file was moved to a `broken` directory, but I guess it's good that it was built in the first place:

```bash
% ls -talh /Users/mickyfitz/micromamba/envs/conda-builder/conda-bld/broken/pam-0.2.5.dev0-py_0.tar.bz2

-rw-r--r--@ 1 mickyfitz  staff   750K 13 Jul 18:56 /Users/mickyfitz/micromamba/envs/conda-builder/conda-bld/broken/pam-0.2.5.dev0-py_0.tar.bz2
```

brynpickering commented:
Thanks for catching that, it was an issue created on a recent PR merge (#205). Now fixed.

This is why I have split out the `conda build` part into its own GH action, so that necessary fixes can be made in a release PR, and the release tag moved until everything works as expected. Then the actual publication of the release on anaconda is a relatively safe GH action.

mfitz commented:
The Docker build appears to be broken in this branch @brynpickering. I've seen a successful build from your other branch this morning, so it definitely seems to be specific to this one.

## In AWS CodeBuild
(I still have the emergency alerts thing switched it off, but it might actually have been helpful here 😄)

<img width="1004" alt="Screen Shot 2023-07-14 at 11 27 28" src="https://github.com/arup-group/pam/assets/250899/2dc4e858-132b-4c62-aefe-41c45cbe9473">


## Locally (same error on both my Intel and M2 machines)

```bash
 => => transferring context: 44.23MB                                                                                                      2.1s
 => [2/4] COPY --chown=mambauser:mambauser . .                                                                                            1.6s
 => ERROR [3/4] RUN micromamba install -y -n base -c city-modelling-lab -f requirements/base.txt &&     micromamba clean --all --yes      1.3s
------
 > [3/4] RUN micromamba install -y -n base -c city-modelling-lab -f requirements/base.txt &&     micromamba clean --all --yes:
#8 0.461
#8 0.461                                            __
#8 0.461           __  ______ ___  ____ _____ ___  / /_  ____ _
#8 0.461          / / / / __ `__ \/ __ `/ __ `__ \/ __ \/ __ `/
#8 0.461         / /_/ / / / / / / /_/ / / / / / / /_/ / /_/ /
#8 0.461        / .___/_/ /_/ /_/\__,_/_/ /_/ /_/_.___/\__,_/
#8 0.461       /_/
#8 0.461
#8 1.319 error    libmamba Could not solve for environment specs
#8 1.319     The following packages are incompatible
#8 1.319     ├─ click <9  does not exist (perhaps a typo or a missing channel);
#8 1.319     ├─ gdal <3.6  does not exist (perhaps a typo or a missing channel);
#8 1.319     ├─ geopandas >=0.13,<0.14  does not exist (perhaps a typo or a missing channel);
#8 1.319     ├─ ipykernel <7  does not exist (perhaps a typo or a missing channel);
#8 1.319     ├─ lxml <5  does not exist (perhaps a typo or a missing channel);
#8 1.319     ├─ matplotlib >=3,<4  does not exist (perhaps a typo or a missing channel);
#8 1.319     ├─ numpy >=1,<2  does not exist (perhaps a typo or a missing channel);
#8 1.319     ├─ pandas >=1.5,<3  does not exist (perhaps a typo or a missing channel);
#8 1.319     ├─ plotly >=4,<6  does not exist (perhaps a typo or a missing channel);
#8 1.319     ├─ prettytable >=3,<4  does not exist (perhaps a typo or a missing channel);
#8 1.319     ├─ python-levenshtein >=0.21,<0.22  does not exist (perhaps a typo or a missing channel);
#8 1.319     ├─ rich >=12,<14  does not exist (perhaps a typo or a missing channel);
#8 1.319     ├─ rtree >=1,<2  does not exist (perhaps a typo or a missing channel);
#8 1.319     ├─ s2sphere <0.3  is uninstallable because it requires
#8 1.319     │  └─ future >=0.15 , which does not exist (perhaps a missing channel);
#8 1.319     ├─ scikit-learn >=1.2,<2  does not exist (perhaps a typo or a missing channel);
#8 1.319     ├─ shapely >=1,<3  does not exist (perhaps a typo or a missing channel);
#8 1.319     └─ xlrd >=2,<3  does not exist (perhaps a typo or a missing channel).
#8 1.319 critical libmamba Could not solve for environment specs
------
executor failed running [/usr/local/bin/_dockerfile_shell.sh micromamba install -y -n base -c city-modelling-lab -f requirements/base.txt &&     micromamba clean --all --yes]: exit code: 1
```

mfitz commented on README.md:
````diff
@@ -46,13 +46,15 @@ supply or utility demand.
 ```
-git clone git@github.com:arup-group/pam.git
-mamba env create -f pam/environment.yml
+mamba create -n pam -c city-modelling-lab pam
````
This command fails for me. I'm guessing this command won't be expected to work until we've published PAM on that channel?

brynpickering replied:
Correct. Could revert this to match the dev docs install instructions until the release PR?

mfitz replied:
It's not a biggie, but on balance I think that's worth it so as not to have the install instructions in a (temporarily) broken state.

mfitz commented on docs/contributing/coding.md:
```diff
@@ -7,7 +7,27 @@ This means we would like help with profiling and implementing parallel compute,
+1. Install mamba with the [Mambaforge](https://github.com/conda-forge/miniforge#mambaforge) executable for your operating system.
+2. Open the command line (or the "miniforge prompt" in Windows).
+3. Download (a.k.a., clone) the PAM respository: `git clone git@github.com:arup-group/pam.git`
+4. Create the PAM mamba environment: `mamba create -n pam -c city-modelling-lab --file requirements/base.txt --file requirements/dev.txt`
```
I had to add the `conda-forge` channel, which I guess is because I haven't globally configured mamba, so for me the command was:

```bash
mamba create -n pam -c city-modelling-lab -c conda-forge --file requirements/base.txt --file requirements/dev.txt
```

If we assume the user is a Mamba n00b like me, maybe it's better to explicitly include the conda-forge channel in the command so it doesn't rely on local config?

brynpickering replied:
Depends on how closely you followed the instructions! Mambaforge uses `conda-forge` as a default channel, so if you didn't already have `mamba` and installed Mambaforge, it should have worked.

mfitz replied:
Ah, I brew installed mamba, which gave me micromamba, which I then aliased to `mamba`.

brynpickering replied:
dangerous, they are two different beasts. Although I am generally in favour of micromamba, it's not the recommended "in" for novices ;)

mfitz commented on docs/contributing/coding.md:
```diff
@@ -7,7 +7,27 @@ This means we would like help with profiling and implementing parallel compute,
+git clone git@github.com:arup-group/pam.git
+mamba create -n pam -c city-modelling-lab --file requirements/base.txt --file requirements/dev.txt
+mamba activate pam
+pip install --no-deps -e ./pam
```
I had to edit this command to `pip install --no-deps -e .` to make it work, otherwise pip couldn't find `pyproject.toml` .

That's because I'd changed into the `pam` directory, where I guess the instructions assume the user is still in the directory above that? Do we want the user in the directory above the pam repo, or in the pam repo? Should we instruct them to `cd pam`?

brynpickering replied:
Yeah, either `cd pam` is in the instructions and we use `pip install --no-deps -e .`, or we leave as is. I guess I always `cd` myself, so no reason to leave it like this.

mfitz commented on pyproject.toml:
```diff
@@ -61,3 +61,52 @@ max-complexity = 10
+    { name = "Theodore Chatziioannou", email = "theodore.chatziioannou@arup.com" },
+    { name = "Bryn Pickering", email = "17178478+brynpickering@users.noreply.github.com" },
+]
+description = "The Population activity Modeller (PAM) is a python API for activity sequence modelling."
```
Missed opportunity to link to one of the Medium blog posts.... 🤣

brynpickering replied:
You may have noticed I slipped a Medium blog badge in the README a few PRs ago, which renders on GitHub at least! It would also render on the PyPI channel if we uploaded PAM there.

mfitz replied:
Actually, I hadn't noticed that - very cool!

mfitz reviewed (approved):
Looks good. I'll need to see the GitHub actions in...action to really grok it, but it makes sense.

I worked through installation and build instructions and found I had to make a couple of slight edits. The conda channel one - I think it's worth being explicit about.

The other one, which was because I'd automatically changed directory even though the instructions didn't tell me to - one could argue that was me failing to follow instructions. But is there any harm in instructing the user to change into the repo they've just cloned?
